Signed release packages using Github artifact attestation

egecetin opened the pull request:
Recently GitHub release public beta for artifact signing https://github.blog/2024-05-02-introducing-artifact-attestations-now-in-public-beta/. With this PR, generated artifacts should also generate `.sigstore.json` files to verify packages. It should be ready to review. Since I changed permissions from a PR, looks like write permissions prohibited. I can't find a way to tweak it to get passing result from CI.

Note: Unfortunately requires drop of Ubuntu 18.04 support because of node version

- All attestations can be found in summaries: https://github.com/egecetin/PcapPlusPlus/actions/runs/9702477673 
- Example attestation page: https://github.com/egecetin/PcapPlusPlus/attestations/1193027
- Example release page: https://github.com/egecetin/PcapPlusPlus/releases/tag/v24.1

egecetin commented:
@seladb I updated this PR and finished work. Only need (or dependency) is #1408. After the merge of #1408 just need to test with Ubuntu 24.04 and it is ready to merge. You can review this PR when you have time.

seladb commented:
@egecetin we removed Ubuntu 18.04, do you think you can update the PR and resolve conflicts?
If the PR is ready for review I can review it

egecetin commented:
@seladb It is ready

tigercosmos commented:
@egecetin  seems the CI have some errors?

egecetin commented:
@tigercosmos It is because of permissions. Since the write permissions set from a PR it prohibits the write permission so, it fails. You can check my master branch or run from my test tag  https://github.com/egecetin/PcapPlusPlus/actions/runs/9858969300

tigercosmos commented:
> @tigercosmos It is because of permissions. Since the write permissions set from a PR it prohibits the write permission so, it fails. You can check my master branch or run from my test tag https://github.com/egecetin/PcapPlusPlus/actions/runs/9858969300

I see. it's cool!

egecetin commented:
@seladb This PR is also ready for review but what do you think about it? Merge before release or after release? It is nice to merge before release but I'm ok for both options since already have many changes for release and last release was long time ago...

clementperon commented on .github/workflows/package.yml:
```diff
@@ -139,6 +157,12 @@ jobs:
         run: cmake --build "$BUILD_DIR" --target package
 
+      - name: Generate artifact attestation
+        id: attestation
```
For my culture why this job need an id ?

egecetin replied:
To be honest, I don't know also. Their example has this structure so i just copied 😁

clementperon replied:
I think you can drop it if you don't use a variable like ${{ steps.attestation.XXXX }}

seladb replied:
yes I agree 👍

seladb commented:
> @seladb This PR is also ready for review but what do you think about it? Merge before release or after release? It is nice to merge before release but I'm ok for both options since already have many changes for release and last release was long time ago...

@egecetin I'm ok with merging this PR before the next release

seladb reviewed (approved):
Fix this comment: https://github.com/seladb/PcapPlusPlus/pull/1390#discussion_r1729467931 and merge!

egecetin commented on CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 # We need 3.12 or later, so that we can set policy CMP0074; see below.
 cmake_minimum_required(VERSION 3.12)
 
-set(PCAPPP_VERSION "23.09+")
```
Oops this for testing from my tags. I should revert it

egecetin replied:
I generated a test tag again. If everything passes I'll revert it

egecetin replied:
Fixed with e5e08cd

egecetin commented:
@seladb last test run: https://github.com/egecetin/PcapPlusPlus/actions/runs/10544926628

seladb commented:
> @seladb last test run: https://github.com/egecetin/PcapPlusPlus/actions/runs/10544926628

Where are the packages created? This tag doesn't have any packages attached... 🤔

egecetin commented:
It was draft. Released now

seladb commented:
> It was draft. Released now
> 
> https://github.com/egecetin/PcapPlusPlus/releases/tag/v24.2

@egecetin I couldn't find the `.sigstore.json` file in the atrifacts 🤔

egecetin commented:
@seladb I removed them because GitHub stores them in attestation tab already (Check [here](https://github.com/egecetin/PcapPlusPlus/attestations)) For verifying there is no need to download `.sigstore.json`. In old commits I added them but then noticed a future problem. GitHub generates them as `jsonl`. This means that every attestation stored in lines of a file. If we add deb/rpm packaging support there will be more than one packages per platform. So we should distinguish sigstore files from lines and to do this we should add a script to parse which line corresponds to which package.

seladb commented on README.md:
````diff
@@ -96,6 +96,20 @@ git clone https://github.com/seladb/PcapPlusPlus.git
 
 Follow the build instructions according to your platform in the [Build From Source](https://pcapplusplus.github.io/docs/install#build-from-source) page in PcapPlusPlus web-site.
 
+### Verify your packages
+
+PcapPlusPlus releases which newer than v23.09 are signed with GitHub attestation. You can verify the attestation of these packages with GitHub CLI. To verify packages you can follow the most recent instructions from [gh attestation verify](https://cli.github.com/manual/gh_attestation_verify). For simple instructions you can use the following command:
+
+```shell
+gh attestation verify <path-to-package-file> --repository seladb/PcapPlusPlus
+```
+
+and you should see the following output in your terminal:
+
+```shell
+✓ Verification succeeded!
+```
````
1. I think it'd be helpful to add a link to the attestation page : https://github.com/seladb/PcapPlusPlus/attestations
2. We should also add it to PcapPlusPlus web-site, maybe here? https://pcapplusplus.github.io/docs/next/install#github-release

seladb reviewed (approved):
Please see 1 comment, otherwise LGTM
